### PR TITLE
Implement GitHub Enterprise License and Organization API's

### LINKED ISSUE #1023: Implement Enterprise - License client
New in GitHub Enterprise 2.4

https://developer.github.com/v3/enterprise/license/

```
/enterprise/license/
 - GET /enterprise/settings/license
```

### LINKED ISSUE #1025: Implement Enterprise - Orgs client
New in Enterprise 2.4

https://developer.github.com/v3/enterprise/orgs/

```
/enterprise/orgs/
 - POST /admin/organizations
```

### PR DESCRIPTION
**Pull request by ryangribble**
## Status
:rocket:Ready to merge:rocket:

## Enhance integration test support for GitHub Enterprise
https://github.com/tattsgroup/octokit.net/commit/535709c3689eaa6b6a4556757c11d0d50b0aae0e
Added a new EnterpriseHelper class for creating enterprise clients, using a new set of environment variables for this purpose.  Enhanced the powershell script to assist in setting up these GitHub Enterprise integration test variables if desired

## Implement Enterprise License Client 
https://developer.github.com/v3/enterprise/license/
Fixes #1023
Includes normal and reactive implementations plus unit and integration tests for both

## Implement Enterprise Organization Client 
https://developer.github.com/v3/enterprise/orgs/
Fixes #1025
Includes normal and reactive implementations plus unit and integration tests for both

## Tests
Obviously the integration tests require a GitHub enterprise instance so wont run as part of Octokit builds currently.  Integration tests are passing on my GHE instance

### CODE REVIEW

**Comment by ryangribble** [comment]
I've rebased this now that the AdminStats PR #1049 has been merged so it should be good for review/merge!

File: Octokit.Tests.Integration/EnterpriseHelper.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Octokit.Tests.Integration
+{
+    public static class EnterpriseHelper
```
Comment (shiftkey) [inline Octokit.Tests.Integration/EnterpriseHelper.cs]: :heart_eyes: this is a great idea

File: Octokit.Tests.Integration/EnterpriseHelper.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+        static EnterpriseHelper()
+        {
+            // Force reading of environment variables.
+            // This wasn't happening if UserName/Organization were 
+            // retrieved before Credentials.
+            Debug.WriteIf(GHECredentials == null, "No credentials specified.");
+        }
+
+        public static string GHEUserName { get; private set; }
```
Comment (shiftkey) [inline Octokit.Tests.Integration/EnterpriseHelper.cs]: :grey_question: rather than the abbreviation, would it be worth just prefixing these properties with `Enterprise` to make this a bit more readable? Or is that redundant given we're already in `EnterpriseHelper`...

File: Octokit.Tests.Integration/Helper.cs [hunk 3]
```diff
@@ -50,9 +50,9 @@ public static class Helper
             return new Credentials(githubUsername, githubPassword);
         });
 
-        static readonly Lazy<Uri> _gitHubEnterpriseUrl = new Lazy<Uri>(() =>
+        static readonly Lazy<Uri> _customUrl = new Lazy<Uri>(() =>
         {
-            string uri = Environment.GetEnvironmentVariable("OCTOKIT_GITHUBENTERPRISEURL");
+            string uri = Environment.GetEnvironmentVariable("OCTOKIT_CUSTOMURL");
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Helper.cs]: :grey_question: I kinda liked having `ENTERPRISE` in this environment variable - could you tell me about what motivated this change?

File: Octokit.Tests.Integration/Helper.cs [hunk 4]
```diff
@@ -80,7 +80,9 @@ static Helper()
 
         public static Credentials BasicAuthCredentials { get { return _basicAuthCredentials.Value; } }
 
-        public static Uri GitHubEnterpriseUrl {  get { return _gitHubEnterpriseUrl.Value; } }
+        public static Uri CustomUrl {  get { return _customUrl.Value; } }
+
+        public static Uri TargetUrl {  get { return CustomUrl ?? GitHubClient.GitHubApiUrl; } }
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Helper.cs]: This is a good bit of :lipstick: :thumbsup:

File: script/configure-integration-tests.ps1 [hunk 5]
```diff
@@ -71,11 +73,26 @@ VerifyEnvironmentVariable "account name" "OCTOKIT_GITHUBUSERNAME"
...
-VerifyEnvironmentVariable "GitHub Enterprise Server URL" "OCTOKIT_GITHUBENTERPRISEURL" $true
+VerifyEnvironmentVariable "Override GitHub URL" "OCTOKIT_CUSTOMURL" $true
 
 VerifyEnvironmentVariable "application ClientID" "OCTOKIT_CLIENTID" $true
-VerifyEnvironmentVariable "application Secret" "OCTOKIT_CLIENTSECRET" $true
\ No newline at end of file
+VerifyEnvironmentVariable "application Secret" "OCTOKIT_CLIENTSECRET" $true
+
+
+if (AskYesNoQuestion "Do you wish to enable GitHub Enterprise (GHE) Integration Tests?" "OCTOKIT_GHE_ENABLED")
```
Comment (shiftkey) [inline script/configure-integration-tests.ps1]: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes:

**Comment by shiftkey** [comment]
@ryangribble this is **so** good!



Just a few conceptual questions about the `EnterpriseHelper` changes, and apologies for introducing the merge conflict on the csproj files!

File: Octokit.Tests.Integration/EnterpriseHelper.cs (hunk 2 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/EnterpriseHelper.cs]: So originally I had this stuff sitting in amongst the existing Helper class and so needed to be very explicit about whether referring to normal or enterprise.  Even when I moved the enterprise stuff out into ```EnterpriseHelper``` I still felt like it was a good thing to be very explicit and obvious that we are dealing with a different set of settings here, than those configured for regular github.

But in hindsight since it's in EnterpriseHelper class anyway, it's probably ok to just call these parmeters UserName Credentials etc rather than GHEUserName GHECredentials and so on.

So I think ill just name these back to their "standard" names

File: Octokit.Tests.Integration/Helper.cs (hunk 3 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/Helper.cs]: So originally that was my "hack" to allow me to run regular integration tests against a GitHub enterprise instance.  But now that I've officially added an EnterpriseHelper class it isnt appropriate to refer to this (in normal helper) as Enterprise URL.  This item still remains so that someone COULD point "all" integration tests at another instance (eg enterprise) if they wanted, by setting the CustomURL.  But the difference is, that all the other paramters like username, password, apitoken etc would be coming from the "regular" settings.

Meanwhile for tests that explicitly require/target GitHub Enterprise, we use the EnterpriseHelper class which has it's own setup for username, password, api key and so on.  

Basically this means that I can have my workstation configured to run integration tests against github.com and enteprise integration tests against my enterprise instance... Plus if I ever need to run the "normal" integration tests against my enterprise instance, I can set the CUSTOM_URL property.

Hope that makes sense

File: Octokit.Tests.Integration/EnterpriseHelper.cs (hunk 2 above)
Comment (shiftkey) [inline Octokit.Tests.Integration/EnterpriseHelper.cs]: :thumbsup:

**Comment by ryangribble** [comment]
Looks like the Travis build failure is unrelated, @shiftkey  ?

**Comment by ryangribble** [comment]
Did the travis build automatically retry or did someone kick it off again?  Is there any way (apart from pushing another commit) for plebs to get the builds to try again?  :grinning:

**Comment by shiftkey** [comment]
> Did the travis build automatically retry or did someone kick it off again? 

Nope, it was me.

>  Is there any way (apart from pushing another commit) for plebs to get the builds to try again?  :grinning:

I'm afraid not. We're looking into the Mono issues and where they might be hiding, but that's going to take some time. I'll keep an eye on the build state as part of reviewing - if you're happy that the error isn't your fault just comment so I know you've looked at it.

**Comment by ryangribble** [comment]
Hey @shiftkey not sure if you may have forgotten about this one, but ive pulled latest master over again and the builds are still passing.  Should be ready for a LGTM :tongue:

File: Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        [GitHubEnterpriseTest]
+        public async Task CanCreateOrganization()
+        {
+            string orgLogin = Helper.MakeNameWithTimestamp("MyOrganization");
+            string orgName = String.Concat(orgLogin, " Display Name");
+
+            var newOrganization = new NewOrganization(orgLogin, EnterpriseHelper.UserName, orgName);
+            var observable = _github.Enterprise.Organization.Create(newOrganization);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs]: Is it worth doing any cleanup here after creating the organization?

File: Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            var organization = await observable;
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs]: :lipstick: as `observable` is only used here you can await the previous line:

`var organization = await _github.Enterprise.Organization.Create(newOrganization);`

File: Octokit.Tests.Integration/Helper.cs (hunk 3 above)
Comment (shiftkey) [inline Octokit.Tests.Integration/Helper.cs]: :thumbsup:

File: Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs (hunk 6 above)
Comment (shiftkey) [inline Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs]: Of course as soon as I realised that I should read the documentation I realise that "there's no endpoint". So that's a pretty good reason to not do that. Nevermind.

**Comment by shiftkey** [comment]
I'm gonna take this in without that bit of :lipstick:. Thanks @ryangribble!

File: Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs (hunk 7 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/Reactive/Enterprise/ObservableEnterpriseOrganizationClientTests.cs]: yeah i think i saw another observable test somewhere that was always getting the observable firs , then awaiting it... so thats how ive written all of my observable tests too
